File: apps/frontend/lib/types/quality.ts
/**
 * Data quality scoring system types (issue #102).
 *
 * Mirrors backend schemas in apps/backend/app/schemas/quality.py and
 * apps/backend/app/schemas/version.py — change both together.
 */

export interface ActionableRecommendation {
  dimension: string
  description: string
  transformation_type: string
  affected_columns: string[]
  severity: 'low' | 'medium' | 'high'
}

export interface QualityGateResult {
  gate_name: string
  passed: boolean
  actual_score: number
  required_score: number
  failing_dimensions: string[]
  is_blocking: boolean
}

export interface QualityReportResponse {
  file_id: string
  filename: string | null
  score_0_100: number
  component_scores: Record<string, number>
  recommendations: string[]
  actionable_recommendations: ActionableRecommendation[]
  gates: QualityGateResult[]
  critical_issue_count: number
  warning_count: number
  partial: boolean
}

export interface QualityTrendPoint {
  version_number: number | null
  created_at: string
  score_before: number | null
  score_after: number | null
  improvement: number | null
  transformation: string
}

export interface QualityTrendResponse {
  dataset_id: string
  points: QualityTrendPoint[]
  overall_improvement: number | null
  best_score: number | null
  worst_score: number | null
}

File: apps/backend/app/schemas/quality.py
"""Schemas for the data quality scoring system (issue #102).

Mirrored by the frontend in apps/frontend/lib/types/quality.ts — change both together.
"""

from pydantic import BaseModel, Field

from app.services.data_processing.quality_assessment import ActionableRecommendation


class QualityGateResult(BaseModel):
    """Result of evaluating one soft quality gate (issue #102, AC4)."""

    gate_name: str
    passed: bool
    actual_score: float = Field(..., description="0-100 overall score evaluated")
    required_score: float = Field(..., description="0-100 minimum overall score")
    failing_dimensions: list[str] = Field(
        default_factory=list, description="Dimensions below their threshold"
    )
    is_blocking: bool = Field(
        False, description="Soft gates never block workflow progression"
    )


class QualityReportResponse(BaseModel):
    """Consolidated quality report for a dataset (issue #102, AC5)."""

    file_id: str
    filename: str | None = None
    score_0_100: float
    component_scores: dict[str, float] = Field(default_factory=dict)
    recommendations: list[str] = Field(default_factory=list)
    actionable_recommendations: list[ActionableRecommendation] = Field(default_factory=list)
    gates: list[QualityGateResult] = Field(default_factory=list)
    critical_issue_count: int = 0
    warning_count: int = 0
    partial: bool = Field(
        False, description="True when computed from a pre-#102 cached report"
    )
